gstaichi/python/dlpack_funcs.cpp

```diff
@@ -0,0 +1,131 @@
+#include "dlpack_funcs.h"
+
+#include "dlpack/dlpack.h"
+
+#include "gstaichi/program/ndarray.h"
+#include "gstaichi/program/program.h"
+#if TI_WITH_CUDA
+#include "gstaichi/rhi/cuda/cuda_device.h"
+#endif  // TI_WITH_CUDA
+#include "gstaichi/rhi/cpu/cpu_device.h"
+
+namespace gstaichi {
+namespace lang {
+pybind11::capsule ndarray_to_dlpack(Program *program,
+                                    pybind11::object owner,
+                                    Ndarray *ndarray) {
+  auto *owner_holder = new pybind11::object(owner);
+  new pybind11::object(owner);
+
+  DeviceAllocation devalloc = ndarray->get_device_allocation();
+
+  void *raw_ptr = nullptr;
+  DLDeviceType device_type = DLDeviceType::kDLCPU;
+
+  cpu::CpuDevice *cpu_device = dynamic_cast<cpu::CpuDevice *>(devalloc.device);
+  if (cpu_device != nullptr) {
+    cpu::CpuDevice::AllocInfo alloc_info = cpu_device->get_alloc_info(devalloc);
+    raw_ptr = alloc_info.ptr;
+  }
+#if TI_WITH_CUDA
+  cuda::CudaDevice *cuda_device =
+      dynamic_cast<cuda::CudaDevice *>(devalloc.device);
```

Comment:
Same, it should at least be an `else if` or inside a `else` clause to avoid trying to cast something that has no chance to work.